pyccl/pkresponse.py

```diff
@@ -0,0 +1,795 @@
+    = \\frac{d\\log P_{mm}(k)}{dh}/(2\\frac{d\\log D}{dh})`.
+
+    Args:
+        cosmo (:class:`~pyccl.core.Cosmology`): a Cosmology object.
+        deltah (float): the variation of h to compute T_{h}(k) by
+            the two-sided numerical derivative method.
+        a_arr (array): an array holding values of the scale factor
+            at which the trispectrum should be calculated for
+            interpolation. If `None`, the internal values used
+            by `cosmo` will be used.
+        lk_arr (array): an array holding values of the natural
+            logarithm of the wavenumber (in units of Mpc^-1) at
+            which the trispectrum should be calculated for
+            interpolation. If `None`, the internal values used
+            by `cosmo` will be used.
+        extrap_order_lok (int): extrapolation order to be used on
+            k-values below the minimum of the splines. See
+            :class:`~pyccl.tk3d.Tk3D`.
+        extrap_order_hik (int): extrapolation order to be used on
+            k-values above the maximum of the splines. See
+            :class:`~pyccl.tk3d.Tk3D`.
+        use_log (bool): if `True`, the trispectrum will be
+            interpolated in log-space (unless negative or
+            zero values are found).
+
+    Returns:
+        Response of the matter power spectrum.
+    """
+
+    if lk_arr is None:
+        status = 0
+        nk = lib.get_pk_spline_nk(cosmo.cosmo)
+        lk_arr, status = lib.get_pk_spline_lk(cosmo.cosmo, nk, status)
+        check(status, cosmo=cosmo)
+    if a_arr is None:
+        status = 0
+        na = lib.get_pk_spline_na(cosmo.cosmo)
+        a_arr, status = lib.get_pk_spline_a(cosmo.cosmo, na, status)
+        check(status, cosmo=cosmo)
+
+    k_use = np.exp(lk_arr)
+
+    # set h-modified cosmology to take finite differencing
+    cosmo_hp, cosmo_hm = set_hmodified_cosmology(cosmo, deltah)
+
+    # Growth factor
+    Dp = cosmo_hp.growth_factor_unnorm(a_arr)
+    Dm = cosmo_hm.growth_factor_unnorm(a_arr)
+
+    # Power spectrum
+    cosmo.compute_linear_power()
+    cosmo_hp.compute_linear_power()
+    cosmo_hm.compute_linear_power()
+
+    pk2dlin = cosmo.get_linear_power("delta_matter:delta_matter")
+
+    pk2d = cosmo.get_nonlin_power("delta_matter:delta_matter")
+    pk2d_hp = cosmo_hp.get_nonlin_power("delta_matter:delta_matter")
+    pk2d_hm = cosmo_hm.get_nonlin_power("delta_matter:delta_matter")
+
+    na = len(a_arr)
+    nk = len(k_use)
+    dpk12 = np.zeros([na, nk])
+    dpk = np.zeros(nk)
+    T_h = np.zeros(nk)
+
+    kmin = 1e-2
+    for ia, aa in enumerate(a_arr):
+        pk = pk2d.__call__(k_use, aa, cosmo)
+        pk_hp = pk2d_hp.__call__(k_use, aa, cosmo_hp)
+        pk_hm = pk2d_hm.__call__(k_use, aa, cosmo_hm)
+
+        dpknl = pk2d.__call__(k_use, aa, cosmo, derivative=True)
+        dpklin = pk2dlin.__call__(k_use, aa, cosmo, derivative=True)
+
+        # use the perturbation theory below kmin
+        T_h[k_use <= kmin] = 1
```

Comment:
This can be moved out of the loop

Reply:
I moved the line (T_h = 1) out of the loop.